From 9f649846dac5ff637fd4e46a76788df2c3eedbea Mon Sep 17 00:00:00 2001
Message-Id: <9f649846dac5ff637fd4e46a76788df2c3eedbea.1481738366.git.ymankad@redhat.com>
In-Reply-To: <be7cc66c94d4b6c77f9f1e674d4783c7d1c3efac.1481738366.git.ymankad@redhat.com>
References: <be7cc66c94d4b6c77f9f1e674d4783c7d1c3efac.1481738366.git.ymankad@redhat.com>
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Tue, 13 Dec 2016 03:04:48 -0500
Subject: [CHANGE 09/10] net/e1000: update network information when macaddr is
 changed in guest
To: ymankad@redhat.com

RH-Author: Vlad Yasevich <vyasevic@redhat.com>
Message-id: <1481598289-10330-4-git-send-email-vyasevic@redhat.com>
Patchwork-id: 73042
O-Subject: [RHEL-6.9 qemu-kvm PATCH 3/4] net/e1000: update network information when macaddr is changed in guest
Bugzilla: 1300626
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Xiao Wang <jasowang@redhat.com>
RH-Acked-by: Laurent Vivier <lvivier@redhat.com>

From: Amos Kong <akong@redhat.com>

If we change macaddr in guest by 'ifconfig eth0 hw ether 12:12:12:34:35:36',
the mac register of e1000 is already updated, but we don't update
network information in qemu. Therefor, the information in monitor
is wrong.

This patch updates nic info when the second part of macaddr is written.

Signed-off-by: Amos Kong <akong@redhat.com>
Signed-off-by: Stefan Hajnoczi <stefanha@redhat.com>
(cherry picked from commit 7c36507c2b8776266f50c5e2739bd18279953b93)

Conflicts:
	hw/net/e1000.c
        - file location.  Lack of qemu_get_queue().

Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Yash Mankad <ymankad@redhat.com>
---
 hw/e1000.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/hw/e1000.c b/hw/e1000.c
index f55d27c..bc06327 100644
--- a/hw/e1000.c
+++ b/hw/e1000.c
@@ -924,7 +924,15 @@ mac_read_clr8(E1000State *s, int index)
 static void
 mac_writereg(E1000State *s, int index, uint32_t val)
 {
+    uint32_t macaddr[2];
+
     s->mac_reg[index] = val;
+
+    if (index == RA + 1) {
+        macaddr[0] = cpu_to_le32(s->mac_reg[RA]);
+        macaddr[1] = cpu_to_le32(s->mac_reg[RA + 1]);
+        qemu_format_nic_info_str(&s->nic->nc, (uint8_t *)macaddr);
+    }
 }
 
 static void
-- 
2.7.4

